File: Graph/Leetcode/Medium/WaysToArriveAtDestination/WaysToArriveAtDestination.cpp
// https://leetcode.com/problems/number-of-ways-to-arrive-at-destination/
#include<bits/stdc++.h>
using namespace std;
int countPaths(int n, vector<vector<int>> &roads)
{
    const int MOD = 1e9 + 7;

    vector<vector<pair<int, int>>> adj(n);

    for (vector<int> &edge : roads)
    {
        adj[edge[0]].emplace_back(make_pair(edge[1], edge[2]));
        adj[edge[1]].emplace_back(make_pair(edge[0], edge[2]));
    }

    vector<long> minCost(n, LONG_MAX);
    minCost[0] = 0;
    vector<long> minWays(n, 0);
    minWays[0] = 1;

    priority_queue<pair<long, long>, vector<pair<long, long>>, greater<pair<long, long>>> pq;

    pq.push(make_pair(0, 0));

    while (!pq.empty())
    {
        pair<long, long> curr = pq.top();
        long cost = curr.first, node = curr.second;
        long ways = minWays[node];
        pq.pop();
        for (pair<int, int> &neighbor : adj[node])
        {
            long newCost = cost + neighbor.second;

            if (minCost[neighbor.first] > newCost)
            {
                minCost[neighbor.first] = newCost;
                minWays[neighbor.first] = ways;
                pq.push(make_pair(newCost, neighbor.first));
            }
            else if (minCost[neighbor.first] == newCost)
            {
                minWays[neighbor.first] = (minWays[neighbor.first] + ways) % MOD;
            }
        }
    }

    return minWays[n - 1] % MOD;
}
int main(void) {
    
    return 0;
}

File: Graph/Leetcode/Medium/WaysToArriveAtDestination/WaysToArriveAtDestination.java
// https://leetcode.com/problems/number-of-ways-to-arrive-at-destination/

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.PriorityQueue;
class Pair {

    long first, second;

    Pair(long first, long second) {
        this.first = first;
        this.second = second;
    }
}
public class WaysToArriveAtDestination {
    public int countPaths(int n, int[][] roads) {
        final int MOD = (int)1e9 + 7;

        List<List<Pair>> adj = new ArrayList<>();
        for(int i=0; i<n; i++) {
            adj.add(new ArrayList<>());
        }

        for(int edge[] : roads) {
            adj.get(edge[0]).add(new Pair(edge[1], edge[2]));
            adj.get(edge[1]).add(new Pair(edge[0], edge[2]));
        }

        long minCost[] = new long[n];
        Arrays.fill(minCost, Long.MAX_VALUE);
        minCost[0] = 0;
        long minWays[] = new long[n];
        minWays[0] = 1;

        PriorityQueue<Pair> pq = new PriorityQueue<>((a, b) -> {
            return Long.compare(a.first,b.first);
        });

        pq.offer(new Pair(0, 0));

        while(!pq.isEmpty()) {
            Pair curr = pq.poll();
            long cost = curr.first, node = curr.second;
            long ways = minWays[(int)node];
            for(Pair neighbor: adj.get((int)node)) {
                long newCost = cost + neighbor.second;
                int adjNode = (int)neighbor.first;
                if(minCost[adjNode] > newCost) {
                    minCost[adjNode] = newCost;
                    minWays[adjNode] = ways;
                    pq.offer(new Pair(newCost, adjNode));
                } else if(minCost[adjNode] == newCost) {
                    minWays[adjNode] = (minWays[adjNode] + ways)%MOD;
                }
            }
        }

        return (int)(minWays[n - 1]%MOD);
    }
}